File: Implement Trie (Prefix Tree).py
#https://leetcode.com/problems/implement-trie-prefix-tree/?envType=study-plan-v2&envId=leetcode-75

class Node:
    def __init__(self):
        self.children={}
        self.endofWord=False
class Trie:

    def __init__(self):
        self.root=Node()
    def insert(self, word: str) -> None:
        curr=self.root

        for w in word:
            if w not in curr.children:
                curr.children[w]=Node()
            curr=curr.children[w]
        curr.endofWord=True

    def search(self, word: str) -> bool:
        curr=self.root

        for w in word:
            if w not in curr.children:
                return False
            curr=curr.children[w]
        if curr.endofWord==True:
            return True
        return False

    def startsWith(self, prefix: str) -> bool:
        curr=self.root
        for w in prefix:
            if w not in curr.children:
                return False
            curr=curr.children[w]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: Implement Trie (Prefix Tree).cpp
//https://leetcode.com/problems/implement-trie-prefix-tree/description/


struct Node{
    Node * child[26];
    bool end=false;
};

class Trie {
private:
    Node* root;
public:
    Trie() {
        root=new Node();
    }
    
    void insert(string word) {
        Node* curr=root;
        for(char c:word){
            int idx=c-'a';
            if(curr->child[idx]==NULL){
                //create new
                Node* node=new Node();
                curr->child[idx]=node;
            }
            curr=curr->child[idx];
        }
        curr->end=true;
    }
    
    bool search(string word) {
        Node* curr=root;
        for(char c:word){
            int idx=c-'a';
            if(curr->child[idx]==NULL){
                return false;
            }
            curr=curr->child[idx];
        }

        return curr->end;
    }
    
    bool startsWith(string prefix) {
        Node* curr=root;
        for(char c:prefix){
            int idx=c-'a';
            if(curr->child[idx]==NULL){
                return false;
            }
            curr=curr->child[idx];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
